script/sync_prod_to_gcp/subscribe_submissions.py

```diff
@@ -0,0 +1,208 @@
+"""
+subscribe_submissions.py is an app that gets the published arxiv ID from the pub/sub queue on GCP,
+check the files exist on CIT and uploads them to GCP bucket.
+
+The "upload to GCP bucket" part is from existing sync_published_to_gcp.
+
+As a matter of fact, this borrows the most of heavy lifting part from sync_published_to_gcp.py.
+"""
+import argparse
+import sys
+import typing
+from time import gmtime
+from pathlib import Path
+
+import json
+import os
+import logging.handlers
+import logging
+
+from google.cloud.pubsub_v1.subscriber.message import Message
+from google.cloud import pubsub_v1, storage
+
+from identifier import Identifier
+from sync_published_to_gcp import ORIG_PREFIX, FTP_PREFIX, upload, ArxivSyncJsonFormatter, \
+    path_to_bucket_key
+
+logging.basicConfig(level=logging.WARNING, format='(%(loglevel)s): (%(timestamp)s) %(message)s')
+logger = logging.getLogger(__file__)
+logger.setLevel(logging.WARNING)
+
+
+LOG_FORMAT_KWARGS = {
+    "fields": {
+        "timestamp": "asctime",
+        "level": "levelname",
+    },
+    "message_field_name": "message",
+    # time.strftime has no %f code "datefmt": "%Y-%m-%dT%H:%M:%S.%fZ%z",
+}
+
+gs_client = storage.Client()
+
+def submission_message_to_payloads(message: Message) -> typing.Tuple[str, typing.List[typing.Tuple[str, str]]]:
+    """
+    Parse the submission_published message, map it to CIT files and returns the list of
+    files to upload to GCP bucket.
+    The schema is
+    https://console.cloud.google.com/cloudpubsub/schema/detail/submission-publication?project=arxiv-production
+    however, this only cares paper_id and version.
+
+    Since upload() looks at the size of bucket object / CIT file to decide copy or not
+    copy, this will attempt to upload the versioned and latest at the same time but the uploading
+    may or may not happen.
+    """
+    try:
+        json_str = message.data.decode('utf-8')
+    except UnicodeDecodeError:
+        logger.error(f"bad data {str(message.message_id)}")
+        return ("",[])
+
+    try:
+        data = json.loads(json_str)
+    except Exception as exc:
+        logger.warning(f"bad({message.message_id}): {json_str[:1024]}")
+        return ("",[])
+
+    paper_id = data.get('paper_id')
+    version = data.get('version')
+    xid = Identifier(f"{paper_id}v{version}" if version else paper_id)
+    archive = ('arxiv' if not xid.is_old_id else xid.archive)
+    logger.info("Processing %s", xid.ids)
+    pairs = []
+
+    if xid.has_version:
```

Comment:
This appears to be the code to figure out what files to move for .abs and src. The .abs file should always be named consistently.

The source may not have a consistent .tar.gz name. (would be nice to have a arxiv-base Metadata.get_src_files() or something) 

Some other possibilities:

{paperidv}.pdf    pdf only submission
{paperidv}.gz       single file submisison
{paperidv}.html.gz    html source submission

I think there is one case in the system where there is a submission with two source files that the admins manually crafted. It would have: {paperidv}.pdf  and {paperidv}.tar.gz 

This was to get a pdf only submission with ancillary files.

Reply:
I was not intending this to be do-all. But, I guess it's not too far to do-all.

Reply:
If it is new/cross/jref just /data/ftp/arxiv/papers/{YYMM}/{paperidv}.* needs to be synced.

If it is a replacement or wdr additionally /data/orig/arxiv/papers/{YYMM}/{paperid}{version-1}.* needs to be synced.

Reply:
@bdc34 I think I did what mentioned here. Please review again.